### PR TITLE
XGLM - Fix Softmax NaNs when using FP16

### LINKED ISSUE #18049: NaN in XGLM Softmax with FP16
### System Info

- `transformers` version: 4.21.0.dev0
- Platform: Linux-5.3.0-1017-x86_64-with-glibc2.27
- Python version: 3.9.13
- Huggingface_hub version: 0.8.1
- PyTorch version (GPU?): 1.12.0+cu102 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: Yes, 4x V100

### Who can help?

@ydshieh @patrickvonplaten 

### Reproduction

Related to the fixes in #17437 most likely.

I am using an example similar to `test_batched_nan_fp16` in `test_modeling_opt.py`, but for an XGLM model. The only difference with that test is the `torch.cuda.amp.autocast` usage, which I found necessary to perform inference (otherwise I would get an error saying "expected scalar type Float but found Half" coming from the forward of XGLM)

```python
import torch
from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer

# Tested with xglm-564M and 7.5B (the second using `infer_auto_device_map` and 
# `load_checkpoint_and_dispatch` from `accelerate`.
tokenizer = AutoTokenizer.from_pretrained("facebook/xglm-564M", padding_side="left")
model = AutoModelForCausalLM.from_pretrained("facebook/xglm-564M", torch_dtype=torch.float16, use_cache=True).cuda()

batch = tokenizer(["Who are you?", "Joe Biden is the president of"], padding=True, return_tensors="pt")

input_ids = batch["input_ids"].cuda()
attention_mask = batch["attention_mask"].cuda()

with torch.no_grad():
    with torch.cuda.amp.autocast():
        outputs = model(input_ids, attention_mask=attention_mask)
        assert not torch.isnan(outputs.logits[0]).any().item() # Raises an AssertionError
```

### Expected behavior

I would expect the model to have normal logits when using FP16. The spotting of this bug was prompted by an issue of garbage generation when doing batching, despite the left padding and a valid attention mask.

### PR DESCRIPTION
**Pull request by gsarti**
# What does this PR do?

Fixes #18049 following the exact same procedure used in #17437. Beside the added test, I also evaluated the fix on my personal use-case and found the behavior of the fixed model to be consistent when performing single or batched generation.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?

## Who can review?

@patil-suraj @ydshieh @patrickvonplaten

### CODE REVIEW

File: src/transformers/models/xglm/modeling_xglm.py
Comment (ydshieh) [inline src/transformers/models/xglm/modeling_xglm.py:240]: We (HF team) have to remember to add this back once Bart takes the same fix.

**Review by ydshieh — APPROVED** [review]
Thank you, @gsarti, LGTM!

**Review by younesbelkada — APPROVED** [review]
LGTM thanks a lot for the fix :rocket: !

File: src/transformers/models/xglm/modeling_xglm.py
Comment (patrickvonplaten) [inline src/transformers/models/xglm/modeling_xglm.py:342]: @stas00 is this operation costly? Wondering how costly such a `max` operation is

**Review by patrickvonplaten — APPROVED** [review]

File: src/transformers/models/xglm/modeling_xglm.py
Comment (stas00) [reply src/transformers/models/xglm/modeling_xglm.py:342]: what are the contenders? at least `max`, `clamp` and `where`, but probably others as well?

```
$ cat clamp-where-max.py
import torch.utils.benchmark as benchmark
import torch

a = torch.tensor(512)
b = torch.tensor(torch.finfo(torch.float16).min)

t0 = benchmark.Timer(
    stmt='torch.clamp(a, b)',
    setup='',
    globals=dict(a=a, b=b),
)

t1 = benchmark.Timer(
    stmt='torch.max(a, b)',
    setup='',
    globals=dict(a=a, b=b),
)

t2 = benchmark.Timer(
    stmt='torch.where(a > b, a, b)',
    setup='',
    globals=dict(a=a, b=b),
)
print(t0.timeit(1000))
print(t1.timeit(1000))
print(t2.timeit(1000))

$ python clamp-where-max.py
<torch.utils.benchmark.utils.common.Measurement object at 0x7f73f6656040>
torch.clamp(a, b)
  2.41 us
  1 measurement, 1000 runs , 1 thread
<torch.utils.benchmark.utils.common.Measurement object at 0x7f73f6656d60>
torch.max(a, b)
  2.38 us
  1 measurement, 1000 runs , 1 thread
<torch.utils.benchmark.utils.common.Measurement object at 0x7f73f6656040>
torch.where(a > b, a, b)
  4.98 us
  1 measurement, 1000 runs , 1 thread
```

so `max` is slightly faster than `clamp`

but ensure to benchmark with the actual dimensions, though it shouldn't make much of a difference I think.

**Comment by patrickvonplaten** [comment]
Sorry for being so late here @gsarti!

**Comment by patrickvonplaten** [comment]
Hey @gsarti - it seems like a test is failing now: 
```
tests/models/xglm/test_modeling_xglm.py::XGLMModelTest::test_xglm_model_past
```

with 

```
 UnboundLocalError: local variable 'dtype_attn_weights' referenced before assignment
```

File: src/transformers/models/xglm/modeling_xglm.py
Comment (ydshieh) [inline src/transformers/models/xglm/modeling_xglm.py:342]: I think we don't need this line.

File: src/transformers/models/xglm/modeling_xglm.py
Comment (ydshieh) [inline src/transformers/models/xglm/modeling_xglm.py:346]: Instead, we can change this part to 

```python
    if attn_weights.dtype == torch.float16:
        attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(attn_weights.dtype)
```

**Comment by ydshieh** [comment]
H

**Comment by ydshieh** [comment]
Hi @gsarti Sorry for being late for this PR. I re-opened it and give some suggestion for a fix to the failing test. Would you like to update this PR after rebasing your working branch on an updated `main` branch?

File: src/transformers/models/xglm/modeling_xglm.py
Comment (younesbelkada) [reply src/transformers/models/xglm/modeling_xglm.py:346]: I think that we should fix the line on OPT too: https://github.com/huggingface/transformers/blob/34be08efcd4d318785b3eac592f27a3d5dd2144b/src/transformers/models/opt/modeling_opt.py#L221 in case `attention_mask` is set to `None` the forward pass will fail as described in https://github.com/huggingface/transformers/pull/18057#issuecomment-1224589697

Comment (younesbelkada) [reply src/transformers/models/xglm/modeling_xglm.py:346]: But I think that the issue has never been reported since `attention_mask` is never `None`: 
https://github.com/huggingface/transformers/blob/34be08efcd4d318785b3eac592f27a3d5dd2144b/src/transformers/models/opt/modeling_opt.py#L629

Comment (ydshieh) [reply src/transformers/models/xglm/modeling_xglm.py:346]: Good catch! Surprisingly, we don't have test failure for OPT due to this.

Comment (ydshieh) [reply src/transformers/models/xglm/modeling_xglm.py:346]: You answered my question before I asked it 😆

**Comment by ydshieh** [comment]
Hi @gsarti I made the necessary change to pass the tests, and pushed to your branch directly. The remaining failing test is irrelevant to this PR, but I will wait until tomorrow to check again, then I will merge.

cc @patrickvonplaten and @younesbelkada

**Comment by younesbelkada** [comment]
Thanks a lot for the fix @ydshieh !!
I think for consistency we should apply the same changes on OPT too, I will take care of that first thing in the morning tomorrow 💪

File: src/transformers/models/xglm/modeling_xglm.py
Comment (younesbelkada) [inline src/transformers/models/xglm/modeling_xglm.py:345]:
```suggestion
            attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(torch.float16)
```
Maybe this is clearer since some people might think that this line will:
1- First assign `attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32)`, therefore `attn_weights.dtype = torch.float32`
2- Then assign `nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32)` to `torch.float32` 
Even though it is not what will happen maybe just in terms of readability/understandability of the code it is better 🤔 ? Wdyt?

Comment (ydshieh) [reply src/transformers/models/xglm/modeling_xglm.py:345]: I am OK for this change as it is minimal and indeed a bit more clear.

**Pull request merged**
